### PR TITLE
Add gradient definition, gradient mark and encoding

### PR DESCRIPTION
**Pull request by domoritz**
This is https://github.com/vega/vega-lite/pull/4970 but on our fork because of https://github.com/vega/vega-lite/issues/5076

### CODE REVIEW

**Comment by kanitw** [comment]
Please rebase too.

File: src/channeldef.ts [hunk 1]
```diff
@@ -70,7 +72,7 @@ export type ValueDefWithCondition<F extends FieldDef<any>, V extends Value = Val
 
 export type StringValueDefWithCondition<F extends Field, T extends Type = StandardType> = ValueDefWithCondition<
   MarkPropFieldDef<F, T>,
-  string | null
+  string | null | Gradient
```
Comment (kanitw) [inline src/channeldef.ts]: Why is this added to all kind of string value def?

File: src/compile/mark/valueref.ts [hunk 2]
```diff
@@ -383,7 +384,7 @@ export function tooltipForEncoding(
 }
 
 export function text(
-  channelDef: ChannelDefWithCondition<FieldDef<string>, string | number | boolean>,
+  channelDef: ChannelDefWithCondition<FieldDef<string>, string | number | boolean | Gradient>,
```
Comment (kanitw) [inline src/compile/mark/valueref.ts]: Text obviously shouldn't accept a gradient as it's value.

**Comment by kanitw** [comment]
(Copied from the previous PR)


Some check lists to help guide you: 

- [ ] Add to encoding
- [ ] Add to marks
- [ ] Add an example of line + area with gradient

**Comment by kanitw** [comment]
Correction, [stroke supports gradient](https://vega.github.io/editor/#/url/vega/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6SgFY6AK1jcAdpAA04KAHdGAExbYwogAw7FESCkbV8WgEx6DUQujVrGM6ltGKlkWCZnoANrC0A2koQoBBhUN6oyFqQjnjIAE6E3D7o8QrBYZAUvqzROJCostx4stHW4ZDsjmpaoeFZjoSseDGwyD7I8K0VDZDchHg8Mv44QQ197Oie-r0TkPDoCQ7ePhkTlYt4qKw+PgC0Cdyo6xuGPo7IS6dnhcWlMuWZ8954rAm+Nxse8YRf87Bfvt0AAzeIJf59QHIQj7djIEHcBLRZ5hAC6qIAvplsRA0W5DGo0pgxpl6pVIvkoHh0OxOpDsrlkKMwOMNsBIKwtABGeSGCgWAAcmL5HK5OHMfKgApwolEItRIU5WgAzFLsloACwqkVgMVa9UysAATm5CrO+tlhq0grNouVOAAbNbZTq5mFLWAAOwusDc426z2C32Cr3m9kOk2+0TmQOR7n6aVa4XupXiv28pMSzVx9PcyVZsCagOpvXxtWF0O5nmavkap2O8MTT3c1yF7lh+1552F8ydxVlvM+wuOlODnnB9uN0st411o3F6sSxP1v3y1EY8K4sD4kDuBC+ZmBMmoyCUmIADwP9PdkDwAE9CFTIBdHtdbx8nM-VBpEAyAF6JNwWggr47S3moxzoI4dSQESNIxDSdLlFAIKMB0tQFFy24QE2eqnueBT3teTzzA+T4xK+VwQh+6BfjEximD0BGMPAVJ4AkeS3oBRxaBxXGnpBGAwTgHLwSS1K0jeqHoT4mFJjhYDbru7joBeR6kuEHJIuhMitAU7AlKUJx1iRl5mXWQzwAA1gAwtwrB6RYOh4dpCS6fpUCdGCChQGZRFmcpBJQBgCTWSybIhKe5HPksVwMiCRwnKJcHEohUnRHhhjIDI8DcGoVLkn0OXgnUA6GBesH+VAV6LNJkBoRhMTYaWhj3lVdXPsRnW+Q1snyRqWV9Pe5gdYeMTdeN85MloLmtX5HHcNZhWMj4eRlWclTUB8DglZRlzvuVWSAgMEVHVp-QgiC7SeTodArnlPhIjEyK1ENFqXddyC3XQbYLCkz0FMoiDMJl80XdwV03Ty911o9gNQHS62KWcm6bZiKNhO9UCsIQ8GFUdsR6YkySpOkKWeNQqwxHEJMpGkYOE2hewAPK2PAzDtSlORrVSZoDtjRjcFQEKiUzjCs+znOwTz604Hd65nJjinKSA2JAA) too, so I think fill/stroke/color all makes sense for gradient.

**Comment by domoritz** [comment]
I made https://github.com/vega/vega-lite/pull/5180 to replace this.
